services/121-service/src/payments/fsp-integration/intersolve-visa/intersolve-visa.api.service.ts

```diff
@@ -66,50 +91,131 @@ export class IntersolveVisaApiService {
+      individual: {
+        firstName: '', // in 121 first name and last name are always combined into 1 "name" field, but Intersolve requires first name, so just give an empty string
+        lastName: input.name,
+        estimatedAnnualPaymentVolumeMajorUnit:
+          input.estimatedAnnualPaymentVolumeMajorUnit,
+      },
+      contactInfo: {
+        addresses: [
+          {
+            type: 'HOME',
+            addressLine1: this.createAddressString(input.contactInformation),
+            city: input.contactInformation.addressCity,
+            postalCode: input.contactInformation.addressPostalCode,
+            country: 'NL',
+          },
+        ],
+        phoneNumbers: [
+          {
+            type: 'MOBILE',
+            value: formatPhoneNumber(input.contactInformation.phoneNumber),
+          },
+        ],
+      },
+    };
+
+    // Send the request
     const authToken = await this.getAuthenticationToken();
     const apiPath = process.env.INTERSOLVE_VISA_PROD
       ? 'customer-payments'
       : 'customer';
     const url = `${intersolveVisaApiUrl}/${apiPath}/v1/customers/create-individual`;
-    console.log('🚀 ~ IntersolveVisaApiService ~ url:', url);
     const headers = [
       { name: 'Authorization', value: `Bearer ${authToken}` },
       { name: 'Tenant-ID', value: process.env.INTERSOLVE_VISA_TENANT_ID },
     ];
-    return await this.httpService.post<IntersolveCreateCustomerResponseBodyDto>(
-      url,
-      payload,
-      headers,
+    const createCustomerResponseDto =
+      await this.httpService.post<CreateCustomerResponseDto>(
+        url,
+        createCustomerRequestDto,
+        headers,
+      );
+
+    // Handle the response
+
+    const errorMessage = this.createErrorMessageIfRequestFailed(
+      createCustomerResponseDto,
     );
+    // If the response contains errors
+    if (errorMessage) {
+      throw new IntersolveVisaApiError(
+        `${IntersolveVisa121ErrorText.createCustomerError}: ${errorMessage}`,
+      );
+    }
+
+    // If the response does not contain errors
+    // Put relevant stuff from createCustomerResponseDto into a CreateCustomerResultDto and return
+    const createCustomerResultDto: CreateCustomerReturnType = {
+      holderId: createCustomerResponseDto.data.data.id,
+    };
+    return createCustomerResultDto;
   }
 
-  public async createWallet(
-    payload: IntersolveCreateWalletDto,
-    brandCode: string,
-  ): Promise<IntersolveCreateWalletResponseDto> {
+  public async issueToken(issueTokenParams: {
+    brandCode: string;
+    activate: boolean;
+    reference?: string;
+  }): Promise<IssueTokenReturnType> {
+    // Create the request body to send
+    const issueTokenRequestDto: IssueTokenRequestDto = {
+      reference: issueTokenParams.reference
+        ? issueTokenParams.reference
+        : uuid(), // A UUID reference which can be used for "technical cancellation in case of time-out", which in accordance with Intersolve we do not implement.
```

Comment:
I would like a bit of explanation, as I do not understand it.

Apart from the syntax:
@PeterSmallenbroek 
Do you know/remember why we put the customer.holderId as reference in case MOCK_INTERSOLVE is true and we let the code generate a UUID if not?
There is probably a good reason for it, but I personally am not in favor of having code that behaves differently based on if it is attached to mock or not. The mock ideally "deals with that".

Reply:
This is indeed to get a specific string/value to the mock service. Because of us needing/wanting to simulate the failure at specific steps on the 'Intersolve' side.  At the moment of writing that, we didn't see another way of getting it to there. Open for suggestions of course.

Reply:
> I would like a bit of explanation, as I do not understand it.

I'm only going to expand on the syntax, not on the logic.

The nullish coalescing operator that I linked above is perfect for this use case: assign something a value only if the first option is [falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy). So, to take an example:

```ts
const reference = issueTokenParams.reference ?? uuid()
```

Is a bit like saying

```ts
let reference = issueTokenParams.reference;
if (reference === false || reference === undefined || reference === null || reference === 'a bunch of other falsy things') {
  reference = uuid();
}
```

With the added value that there are internal optimizations to make it avoid doing all these checks.

Reply:
Code syntax change applied in new branch.

TODO: see if there is a better way to pass data to the mock.

Reply:
The only clean way I can think is to let the test code configure the Mock Service as to behave in the way it wants. In this case: let the call to create a token fail.

In that way we do not need our production code to contain flags and paths for behaving differently when run with a Mock.

However, if there are many tests running against a single Mock Service simultaneously (which is the case now I think), with different needs for Mock configuration, then I do not see a way for the Mock Service to know which configuration belongs to which test.

Maybe this can be overcome with a preview environment as in:
https://dev.azure.com/redcrossnl/121%20Platform/_workitems/edit/28089
Or maybe the automated tests running on GH already are "isolated" including the Mock Service?

@aberonni @PeterSmallenbroek  @RubenGeo What do you think?

Reply:
In principle our API tests do not run in parallel. So you could set the mock-service to certain state (like always fail customer creation) with an API test and than clean up that state at the end of an API test. However the mock-service does not have a DB so this state needs to live in memory. I do think setting and cleaning up such a 'state' is something that is very prone to bugs in the tests, for example if you do not cleanup the state properly you would get failed runs in tests that are not actually broken. (But it could definitely work)

An alternative solution: maybe instead of manipulating the reference (which is an actual parameter of the intersolve API) we could send a mock specific parameter. Like MOCK_SCENARIO='fail-create-customer' in the headers. 

So something like, if our 121-service has been started up with with a certain node_env value set, it forwards all headers of API calls it received. So than our API test that is supposed to let the creation of a customer fail sends all it's request with the  MOCK_SCENARIO='fail-create-customer' in the headers which is than globally set with some middleware) in the 121-service and forwarded to the mock-service with API calls the 121-service makes (also with some middleware). This would have to advantage that mock code runs in a more separated places in the code.

Both of these solutions (the state one and the headers one) have as drawback that they do not consider anything besides API test. You would also want to have the mock scenarios working when manual testing, or when giving a demo. @diderikvw

Reply:
Thanks @RubenGeo ! Clear and viable alternatives.
After some thinking and chatting with AI, I feel that "passing on" HTTP headers is less clean and more complex in the long run. And that "configuring the Mock Service how I want it to behave before starting the test" is more straight-forward. 

For the user of the Mock Service in principle it should not matter if some service has DB or not: that is encapsulated.

To prevent that the Mock Service was not "cleaned up after use", perhaps a practice of "always configure the Mock Service before starting a test" would be a way? Then it does not matter if someone else cleaned up or not.

Reply:
How would you configure the Mock Service? Via an API call?

Could be that you have a CRUD endpoint `/configuration` with an object. Which is:
`{
  mock-intersolve-wallet-creation-fail: true
  mock-intersolve-customer-creation-fail: false
  mock-safaricom-topup: false
 etc...
}`

And than before each test you make an api call to '/configuration?reset=true' or something to reset it. 

You would need to refactor a lot of tests tests and mock service than. Some tests import multiple PA and expect a part of them to fail for example, which is more difficult in this scenario





`

Reply:
@RubenGeo  Thank you for your response! Let's have a conversation.